luminoth/tools/dataset/readers/object_detection/csv_reader.py

```diff
@@ -0,0 +1,226 @@
+import os
+import tensorflow as tf
+import csv
+
+from PIL import Image
+
+from luminoth.tools.dataset.readers import InvalidDataDirectory
+from luminoth.tools.dataset.readers.object_detection import (
+    ObjectDetectionReader
+)
+from luminoth.utils.config import is_basestring
+from luminoth.utils.dataset import read_image
+
+
+# We automatically map common key names to known values.
+# Feel free to add new values commonly found in the wild.
+FIELD_MAPPER = {
+    'x_min': 'xmin',
+    'x1': 'xmin',
+    'x_max': 'xmax',
+    'x2': 'xmax',
+    'y_min': 'ymin',
+    'y1': 'ymin',
+    'y_max': 'ymax',
+    'y2': 'ymax',
+    'class_id': 'label',
+    'class': 'label',
+    'label_id': 'label',
+    'img_name': 'image_id',
+    'img_id': 'image_id',
+    'image': 'image_id',
+}
+
+DEFAULT_COLUMNS = [
+    'image_id', 'xmin', 'ymin', 'xmax', 'ymax', 'label'
+]
+
+
+class CSVReader(ObjectDetectionReader):
+    def __init__(self, data_dir, split, columns=DEFAULT_COLUMNS,
+                 with_header=False, **kwargs):
+        super(CSVReader, self).__init__()
+        self._data_dir = data_dir
+        self._split = split
+        self._labels_filename = self._get_labels_filename()
+
+        if is_basestring(columns):
+            columns = columns.split(',')
+        self._columns = columns
+        self._with_header = with_header
+
+        self._total = None
+        self._classes = None
+        self._files = None
+
+        self.errors = 0
+        self.yielded_records = 0
+
+    @property
+    def total(self):
+        if self._total is None:
+            self._total = len(self._get_records())
+        return self._total
+
+    @property
+    def classes(self):
+        if self._classes is None:
+            self._classes = sorted(set([
+                g['label']
+                for r in self._get_records().values()
+                for g in r
+            ]))
+        return self._classes
+
+    @property
+    def files(self):
+        if self._files is None:
+            files_dir = os.path.join(self._data_dir, self._split)
+            self._files = tf.gfile.ListDirectory(files_dir)
+        return self._files
+
+    def iterate(self):
+        records = self._get_records()
+        for image_id, image_data in records.items():
+
+            image_path = self._get_image_path(image_id)
+            if image_path is None:
+                tf.logging.debug(
+                    'Could not find image_path for image "{}".'.format(
+                        image_id
+                    ))
+                self.errors += 1
+                continue
+
+            try:
+                image = read_image(image_path)
+            except tf.errors.NotFoundError:
+                tf.logging.debug('Could not find image "{}" in "{}".'.format(
+                    image_id, image_path
+                ))
+                self.errors += 1
+                continue
+
+            image_pil = Image.open(image_path)
```

Comment:
Why safer?

Reply:
Because you make sure the file is closed no matter what. Also see [this](https://stackoverflow.com/questions/8774830/how-with-is-better-than-try-catch-to-open-a-file-in-python).